File: merge_klist.py
from Queue import PriorityQueue

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        #O(nklogk)
        #O(k)
        #insert tuple of value, linklist reference
        # head = point = ListNode(0)
        # q = PriorityQueue()
        # for l in lists:
        #     if l:
        #         q.put((l.val, l))
        # while not q.empty():
        #     val, node = q.get()
        #     point.next = ListNode(val)
        #     point = point.next
        #     node = node.next
        #     if node:
        #         q.put((node.val, node))
        # return head.next
        
        
        #while inserting maintain sort order for node values
        #custom sort
        setattr(ListNode, "__lt__", lambda self, other: self.val <= other.val)
            
        pq = []
        for l in lists:
            if l:
                heapq.heappush(pq,  l)
        
        out = ListNode(None)
        head = out
        while pq:
            l = heapq.heappop(pq)
            head.next = l
            head = head.next
            if l and l.next:
                heapq.heappush( pq, l.next)
            
        return out.next

File: mergeKlist.java

//Time complexity : O(N\log N)O(NlogN) 
//Space complexity : O(N)O(N)
class Solution {
    ListNode head;

    public void inserrt(int data)
    {
         ListNode new_node = new ListNode(data);
            if (head == null)
            {
            head = new ListNode(data);
                return;
            }
 
           new_node.next = null;
            ListNode last = head;
            while (last.next != null)
                last = last.next;
            last.next = new_node;
            return;
    }
        
    
    public ListNode mergeKLists(ListNode[] lists) {
        ArrayList<Integer> arr = new ArrayList<>();
        ListNode head1 = null;

        for(int i=0; i<lists.length; i++){
            ListNode head2 = lists[i];
            while(head2 != null){
                arr.add(head2.val);
                head2  = head2.next;
            }
            
        }
        System.out.print(arr);
        Collections.sort(arr);
        System.out.print(arr);

        for(int j=0; j < arr.size(); j++){
               inserrt(arr.get(j));  
        }
        return head;
        
    }
}